dipy/tracking/streamlinespeed.pyx

```diff
@@ -98,16 +98,23 @@ def length(streamlines):
             return 0.0
 
         arclengths = np.zeros(len(streamlines), dtype=np.float64)
-
-        if as_native_array(streamlines._data).dtype == np.float32:
+        streamlines_data = as_native_array(streamlines._data)
+        if streamlines_data.dtype == np.float32:
             c_arclengths_from_arraysequence[float2d](
-                                    as_native_array(streamlines._data),
+                                    streamlines_data,
                                     streamlines._offsets.astype(np.intp),
                                     streamlines._lengths.astype(np.intp),
                                     arclengths)
+        elif streamlines_data.dtype == np.float64:
+            c_arclengths_from_arraysequence[double2d](
+                                      streamlines_data,
+                                      streamlines._offsets.astype(np.intp),
+                                      streamlines._lengths.astype(np.intp),
+                                      arclengths)
         else:
+            streamlines_data = streamlines_data.astype(np.float64)
```

Comment:
So for `np.float16` you effectively convert to `np.float64` I guess because there is no other equivalent type in Cython? Then where is the inverse conversion made when writing? i.e. referring to this comment in the commit message:
```
Saving streamlines as np.float16 can be useful for saving disk space.
```

Reply:
you are right, I should revert back to the original dtype. Let me look at it.

Reply:
thank you for the feedback